subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ZipHasher.java

```diff
@@ -32,82 +34,99 @@
-        return hashJarContents(fileSnapshot);
+        return hashZipContents(fileSnapshot);
     }
 
     @Override
     public void appendConfigurationToHasher(Hasher hasher) {
         hasher.putString(getClass().getName());
-        classpathResourceHasher.appendConfigurationToHasher(hasher);
-        classpathResourceFilter.appendConfigurationToHasher(hasher);
+        resourceHasher.appendConfigurationToHasher(hasher);
+        resourceFilter.appendConfigurationToHasher(hasher);
     }
 
     @Nullable
-    private HashCode hashJarContents(RegularFileSnapshot jarFileSnapshot) {
+    private HashCode hashZipContents(RegularFileSnapshot zipFileSnapshot) {
         try {
-            List<FileSystemLocationFingerprint> fingerprints = fingerprintZipEntries(jarFileSnapshot.getAbsolutePath());
+            List<FileSystemLocationFingerprint> fingerprints = fingerprintZipEntries(zipFileSnapshot.getAbsolutePath());
             if (fingerprints.isEmpty()) {
                 return null;
             }
             Hasher hasher = Hashing.newHasher();
             NormalizedPathFingerprintCompareStrategy.appendSortedToHasher(hasher, fingerprints);
             return hasher.hash();
         } catch (Exception e) {
-            return hashMalformedZip(jarFileSnapshot, e);
+            return hashMalformedZip(zipFileSnapshot, e);
         }
     }
 
-    private List<FileSystemLocationFingerprint> fingerprintZipEntries(String jarFile) throws IOException {
-        List<FileSystemLocationFingerprint> fingerprints = Lists.newArrayList();
-        InputStream fileInputStream = null;
+    private List<FileSystemLocationFingerprint> fingerprintZipEntries(String zipFile) throws IOException {
+        ZipInput input = null;
         try {
-            fileInputStream = Files.newInputStream(Paths.get(jarFile));
-            ZipInputStream zipInput = new ZipInputStream(fileInputStream);
-            ZipEntry zipEntry;
-            RelativePathFactory relativePathFactory = new RelativePathFactory();
-
-            while ((zipEntry = zipInput.getNextEntry()) != null) {
-                relativePathFactory.setZipEntry(zipEntry);
-                if (zipEntry.isDirectory() || classpathResourceFilter.shouldBeIgnored(relativePathFactory)) {
-                    continue;
-                }
-                HashCode hash = classpathResourceHasher.hash(zipEntry, zipInput);
+            input = ZipInput.create(new File(zipFile));
+            return fingerprintZipEntries(input);
+        } finally {
+            IoActions.closeQuietly(input);
+        }
+    }
+
+    private List<FileSystemLocationFingerprint> fingerprintZipEntries(ZipInput input) throws IOException {
+        List<FileSystemLocationFingerprint> fingerprints = Lists.newArrayList();
+        ZipEntry zipEntry;
+        RelativePathFactory relativePathFactory = new RelativePathFactory();
+
+        while ((zipEntry = input.getNextEntry()) != null) {
```

Comment:
Some coding standards prohibit assignment and value checking in the same expression.
Some only use for loops for indexing.
Here's an alternative, if you think it's more compact (i.e. iteration is separated to a single line, even if it contains repetition):
```java
for (ZipEntry zipEntry = input.getNextEntry(); zipEntry != null; zipEntry = input.getNextEntry()) {
```